boa/src/syntax/lexer/identifier.rs

```diff
@@ -0,0 +1,71 @@
+use super::{Cursor, Error, Tokenizer};
+use crate::syntax::ast::{Position, Span};
+use crate::syntax::lexer::token::Numeric;
+use crate::syntax::lexer::{Token, TokenKind};
+use std::io::Read;
+use std::str::FromStr;
+
+/// Identifier or keyword lexing.
+///
+/// This currently includes boolean/NaN lexing.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]:
+/// [mdn]:
+#[derive(Debug, Clone, Copy)]
+pub(super) struct Identifier {
+    init: char,
+}
+
+impl Identifier {
+    /// Creates a new identifier/keyword lexer.
+    pub(super) fn new(init: char) -> Self {
+        Self { init }
+    }
+}
+
+impl<R> Tokenizer<R> for Identifier {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        let mut buf = self.init.to_string();
+
+        loop {
+            match cursor.peek() {
```

Comment:
We should use a `while let Some(c) = cursor.peek() {`.
Then, we can just do `let c = c?;` and forget about errors.

Reply:
Because the error returned is a reference (&) the ? notation fails.

Reply:
> We should use a `while let Some(c) = cursor.peek() {`.
> Then, we can just do `let c = c?;` and forget about errors.

Because .peek() returns a & reference to the error the ? notation fails (I do think if possible we should try and make peek return Result<&Token, Error> so ? can be used).

Reply:
Now that .peek() returns an Error rather than &Error this change has been made.